media-source/mediasource-endofstream.html

```diff
@@ -23,51 +23,54 @@
             mediaSource.endOfStream();
             test.expectEvent(mediaElement, 'canplaythrough',
               'Media element may render the media content until the end');
-            test.waitForExpectedEvents(function()
-            {
-                assert_equals(mediaElement.readyState, HTMLMediaElement.HAVE_ENOUGH_DATA,
-                  'Media element has enough data to render the content');
-                test.done();
-            });
         });
-    }, 'MediaSource.endOfStream(): media element notified that it now has all of the media data');
 
-    mediasource_testafterdataloaded(function(test, mediaElement, mediaSource, segmentInfo, sourceBuffer, mediaData)
-    {
-        sourceBuffer.appendBuffer(mediaData);
-        test.expectEvent(sourceBuffer, 'updateend',
-          'Media buffer appended to SourceBuffer');
         test.waitForExpectedEvents(function()
         {
-            assert_equals(sourceBuffer.buffered.length, 1,
-              'Media data properly buffered');
-            var highestEndTime = sourceBuffer.buffered.end(0);
-
-            mediaSource.endOfStream();
-
-            assert_equals(sourceBuffer.buffered.end(0), highestEndTime,
-              'Ending the stream does not affect buffered data');
+            assert_equals(mediaElement.readyState, HTMLMediaElement.HAVE_ENOUGH_DATA,
+              'Media element has enough data to render the content');
             test.done();
         });
-    }, 'MediaSource.endOfStream(): buffered data not modified when endOfStream is called');
+    }, 'MediaSource.endOfStream(): media element notified that it now has all of the media data');
+
+    function threeDecimalPlaces(number)
+    {
+        return Number(number.toFixed(3));
+    }
 
     mediasource_testafterdataloaded(function(test, mediaElement, mediaSource, segmentInfo, sourceBuffer, mediaData)
     {
-        assert_less_than(segmentInfo.duration, 60, 'Sufficient test media duration');
         sourceBuffer.appendBuffer(mediaData);
         test.expectEvent(sourceBuffer, 'updateend',
           'Media buffer appended to SourceBuffer');
         test.waitForExpectedEvents(function()
         {
             assert_equals(sourceBuffer.buffered.length, 1,
               'Media data properly buffered');
+            var highestEndTime = sourceBuffer.buffered.end(0);
+
+            // Note that segmentInfo.duration is expected to also be the
+            // highest track buffer range end time. Therefore, endOfStream() should
+            // not change duration with this media.
+            assert_equals(threeDecimalPlaces(segmentInfo.duration), threeDecimalPlaces(mediaSource.duration));
+            assert_less_than_equal(highestEndTime, mediaSource.duration,
+                'Media duration may be slightly longer than intersected track buffered ranges');
 
-            mediaSource.duration = 60;
+            // Set the duration even higher, then confirm that endOfStream() drops it back to be
```

Comment:
This is the current editor's draft text in the relevant portion of the end of stream algorithm (as we discussed in the related change: https://github.com/w3c/media-source/pull/154#issuecomment-241078008):

> If error is not set
> Run the duration change algorithm with new duration set to the largest track buffer ranges end time across all the track buffers across all SourceBuffer objects in sourceBuffers.
> NOTE
> This allows the duration to properly reflect the end of the appended media segments. For example, if the duration was explicitly set to 10 seconds and only media segments for 0 to 5 seconds were appended before endOfStream() was called, then the duration will get updated to 5 seconds.